pkg/oc/generate/app/cmd/resolve.go

```diff
@@ -184,14 +187,25 @@ func Resolve(appConfig *AppConfig) (*ResolvedComponents, error) {
 
 // AddSourceRepositoriesToRefBuilder adds the provided repositories to the reference builder, identifies which
 // should be built using Docker, and then returns the full list of source repositories.
-func AddSourceRepositoriesToRefBuilder(b *app.ReferenceBuilder, repos []string, g *GenerationInputs) (app.SourceRepositories, error) {
+func AddSourceRepositoriesToRefBuilder(b *app.ReferenceBuilder, c *ComponentInputs, g *GenerationInputs, s, i *[]string) (app.SourceRepositories, error) {
 	strategy := g.Strategy
 	if strategy == generate.StrategyUnspecified {
 		strategy = generate.StrategySource
 	}
-	for _, s := range repos {
-		if repo, ok := b.AddSourceRepository(s, strategy); ok {
-			repo.SetContextDir(g.ContextDir)
+	// when git is installed we keep default logic. sourcelookup.go will do sorting of images, repos
+	if git.IsGitInstalled() || len(c.SourceRepositories) > 0 {
+		for _, s := range c.SourceRepositories {
+			if repo, ok := b.AddSourceRepository(s, strategy); ok {
+				repo.SetContextDir(g.ContextDir)
+			}
+		}
+		// when git is not installed we need to parse some logic to decide if we got 'new-app -i image code' or 'image -c code' syntax
+	} else if len(c.Components) > 0 && len(*i) > 0 && len(*s) == 0 || len(c.Components) > 0 && len(*i) == 0 && len(*s) > 0 {
```

Comment:
This covers the case when half of the components are passed via flags, and half via Components. We need this because of the use-cases like "image image code" or "image image" where componentInputs becomes complex.